### PR TITLE
Support AWS OpenSearch Serverless

### PR DESCRIPTION
**Pull request by aochsner**
skipping _refresh call as it is not supported

## Hey, I just made a Pull Request!

Continuation on #17619 to disable `_refresh` endpoint when configured for AWS OpenSearch Serverless which doesn't support that endpoint.  

https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-genref.html#serverless-operations

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/search-backend-module-elasticsearch/api-report.md [hunk 1]
```diff
@@ -370,6 +370,7 @@ export type ElasticSearchSearchEngineIndexerOptions = {
   logger: Logger | LoggerService;
   elasticSearchClientWrapper: ElasticSearchClientWrapper;
   batchSize: number;
+  skipRefresh: boolean;
```
Comment (cal5barton) [inline plugins/search-backend-module-elasticsearch/api-report.md:373]: As per @freben in the discord chat.

``` 
freben — Today at 12:18 PM
Hm could skipRefresh be optional? Makes the change non-breaking
And then checking options.skipRefresh !== true at the last lines
```

Comment (aochsner) [reply plugins/search-backend-module-elasticsearch/api-report.md]: makes sense.  hadn't checked there in a while.  Thanks.  Updated!

**Review by tikabom — APPROVED** [review]
Apologies for the delay on this! lgtm

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts [hunk 2]
```diff
@@ -95,7 +96,7 @@ export class ElasticSearchSearchEngineIndexer extends BatchSearchEngineIndexer {
           index: { _index: that.indexName },
         };
       },
-      refreshOnCompletion: that.indexName,
+      refreshOnCompletion: !options.skipRefresh && that.indexName,
```
Comment (tikabom) [inline plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts:99]: if options.skipRefresh is `true` here, `refreshOnCompletion` will be set to `false`. Are we sure that's alright? Or do we want to handle this differently? Maybe we don't want to pass `refreshOnCompletion` instead?

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts [hunk 3]
```diff
@@ -473,6 +481,8 @@ export class ElasticSearchSearchEngine implements SearchEngine {
       return {
         provider: 'aws',
         node: config.getString('node'),
+        region: config.getOptionalString('region'),
```
Comment (freben) [inline plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts:484]: This one should be added to [the config schema](https://github.com/backstage/backstage/blob/b8ce15b1e5bfa2f5184dc61c96af6c45addcd1aa/plugins/search-backend-module-elasticsearch/config.d.ts#L117). I see it was read below already, but while we're here we could update the schema to be appropriate.

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts (hunk 2 above)
Comment (freben) [reply plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts:99]: Also, does this really need to be on the indexer at all? Would it be possible to leave the indexer completely unchanged and instead leverage the fact that the client is wrapped up and segregates opensearch already?

https://github.com/backstage/backstage/blob/b8ce15b1e5bfa2f5184dc61c96af6c45addcd1aa/plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchClientWrapper.ts#L122

Could it just exclude the refreshOnCompletion field in one of the paths there selectively?

Comment (aochsner) [reply plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts]: > if `options.skipRefresh` is `true` here, `refreshOnCompletion` will be set to `false`. Are we sure that's alright? Or do we want to handle this differently? Maybe we don't want to pass `refreshOnCompletion` instead?

It is an optional field but the default is `false` so net effect is the same... https://opensearch.org/docs/latest/api-reference/document-apis/bulk/

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts (hunk 3 above)
Comment (aochsner) [reply plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts:484]: yep can do

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts (hunk 2 above)
Comment (aochsner) [reply plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts]: > Also, does this really need to be on the indexer at all? Would it be possible to leave the indexer completely unchanged and instead leverage the fact that the client is wrapped up and segregates opensearch already?
> 
> https://github.com/backstage/backstage/blob/b8ce15b1e5bfa2f5184dc61c96af6c45addcd1aa/plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchClientWrapper.ts#L122
> 
> Could it just exclude the refreshOnCompletion field in one of the paths there selectively?

I don't think so.  OpenSearch supports refresh.  However, AWS OpenSearch Serverless does not.

Comment (aochsner) [reply plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngineIndexer.ts]: Ready for another round of review.

**Comment by aochsner** [comment]
Bump @freben

**Comment by freben** [comment]
yepyep working through the big pile of reviews now :)

**Review by freben — APPROVED** [review]
Let's go!

**Pull request merged**